File: src/research_artifact_uploader/archive.py
from __future__ import annotations

from datetime import datetime
from pathlib import Path
import hashlib
import tarfile
import json
import yaml
import os

from .collect import CollectedArtifact
from .manifest import Manifest
from .paths import DEFAULT_EXCLUDED_DIRS


def make_timestamp() -> str:
    return datetime.now().strftime("%Y%m%d_%H%M%S")


def run_id_for_run(run_name: str, timestamp: str | None = None) -> str:
    return f"{run_name}_{timestamp or make_timestamp()}"


def compute_sha256(path: Path) -> tuple[str, int]:
    h = hashlib.sha256()
    size = 0
    with path.open("rb") as f:
        while True:
            chunk = f.read(1024 * 1024)
            if not chunk:
                break
            h.update(chunk)
            size += len(chunk)
    return h.hexdigest(), size


def _unique_name(base_name: str, used: set[str]) -> str:
    if base_name not in used:
        used.add(base_name)
        return base_name

    stem = Path(base_name).stem
    suffix = Path(base_name).suffix
    i = 1
    while True:
        candidate = f"{stem}_{i}{suffix}"
        if candidate not in used:
            used.add(candidate)
            return candidate
        i += 1


def _iter_archive_members(
    artifact: CollectedArtifact,
    run_name: str,
    project_root: Path,
    used_dest_names: set[str],
) -> list[tuple[Path, str]]:
    entries: list[tuple[Path, str]] = []

    def _add_file(file_path: Path, arc_dir: Path) -> None:
        base = file_path.name
        final = _unique_name(base, used_dest_names)
        entries.append((file_path, str((arc_dir / final).as_posix())))

    if artifact.type == "file":
        if not artifact.matched_paths:
            return entries
        src = Path(artifact.matched_paths[0])
        arc_dir = Path(run_name) / "artifacts" / artifact.name
        _add_file(src, arc_dir)
        return entries

    if artifact.type == "directory":
        src = Path(artifact.matched_paths[0]) if artifact.matched_paths else None
        if src is None or not src.exists():
            return entries
        for child in sorted(src.rglob("*")):
            if child.is_dir():
                continue
            if any(p in child.as_posix().split(os.sep) for p in DEFAULT_EXCLUDED_DIRS):
                continue
            rel = child.relative_to(src)
            dst = Path(run_name) / "artifacts" / artifact.name / rel
            entries.append((child, str(dst.as_posix())))
        return entries

    if artifact.type == "glob":
        for match_str in artifact.matched_paths:
            match = Path(match_str)
            if match.is_file():
                arc_dir = Path(run_name) / "artifacts" / artifact.name
                base = match.name
                final = _unique_name(base, used_dest_names)
                entries.append((match, str((arc_dir / final).as_posix())))
            elif match.is_dir():
                group = _unique_name(match.name, used_dest_names)
                for child in sorted(match.rglob("*")):
                    if child.is_dir():
                        continue
                    if any(p in child.as_posix().split(os.sep) for p in DEFAULT_EXCLUDED_DIRS):
                        continue
                    rel = child.relative_to(match)
                    dst = Path(run_name) / "artifacts" / artifact.name / group / rel
                    entries.append((child, str(dst.as_posix())))
        return entries

    return entries


def create_archive(
    manifest: Manifest,
    artifacts: list[CollectedArtifact],
    run_id: str,
    run_name: str,
    output_dir: Path,
    manifest_yaml_path: Path | None,
    git_info_path: Path | None,
    metadata_path: Path | None,
) -> Path:
    output_dir.mkdir(parents=True, exist_ok=True)
    archive_path = output_dir / f"{run_id}.tar.gz"

    used_dest_names: set[str] = set()
    with tarfile.open(archive_path, "w:gz", compresslevel=manifest.archive.compression_level) as tar:
        for artifact in artifacts:
            if artifact.status != "included":
                continue
            for source, arc_name in _iter_archive_members(artifact, run_name, Path.cwd(), used_dest_names):
                if source.exists() and source.is_file():
                    tar.add(source, arcname=arc_name, recursive=False)

        if manifest.archive.include_manifest and manifest_yaml_path is not None:
            tar.add(manifest_yaml_path, arcname=f"{run_name}/manifest.yaml", recursive=False)
        if manifest.archive.include_git_info and git_info_path is not None:
            tar.add(git_info_path, arcname=f"{run_name}/git_info.json", recursive=False)
        if manifest.archive.include_metadata and metadata_path is not None and metadata_path.exists():
            tar.add(metadata_path, arcname=f"{run_name}/metadata.json", recursive=False)

    return archive_path


def write_manifest_snapshot(manifest: Manifest, path: Path) -> None:
    path.write_text(yaml.safe_dump(manifest.to_dict(), sort_keys=False), encoding="utf-8")


def write_git_snapshot(payload: dict, path: Path) -> None:
    path.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")

File: src/archive.rs
use std::{
    collections::HashSet,
    fs::{self, File},
    io::{Read, Write},
    path::{Path, PathBuf},
};

use flate2::{write::GzEncoder, Compression};
use sha2::{Digest, Sha256};
use tar::Builder;
use walkdir::WalkDir;

use crate::{
    collect::{CollectedArtifact, ArtifactStatus},
    manifest::Manifest,
    paths::DEFAULT_EXCLUDED_DIRS,
    AppResult,
};

pub fn make_timestamp() -> String {
    chrono::Local::now().format("%Y%m%d_%H%M%S").to_string()
}

pub fn run_id_for_run(run_name: &str, timestamp: Option<&str>) -> String {
    match timestamp {
        Some(v) => format!("{run_name}_{v}"),
        None => format!("{run_name}_{}", make_timestamp()),
    }
}

pub fn compute_sha256(path: &Path) -> AppResult<(String, u64)> {
    let mut file = File::open(path).map_err(|e| format!("open archive failed: {e}"))?;
    let mut hasher = Sha256::new();
    let mut buf = [0u8; 1024 * 1024];
    let mut total = 0u64;

    loop {
        let n = file.read(&mut buf).map_err(|e| format!("read archive failed: {e}"))?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
        total += n as u64;
    }

    Ok((format!("{:x}", hasher.finalize()), total))
}

pub fn write_sha256(archive: &Path, sha256: &str) -> AppResult<PathBuf> {
    let sha_path = archive.with_extension("sha256");
    let mut fh = File::create(&sha_path).map_err(|e| format!("write sha256 failed: {e}"))?;
    writeln!(
        fh,
        "{}  {}",
        sha256,
        archive.file_name().and_then(|n| n.to_str()).unwrap_or("archive")
    )
    .map_err(|e| format!("write sha256 failed: {e}"))?;
    Ok(sha_path)
}

pub fn create_archive(
    manifest: &Manifest,
    artifacts: &[CollectedArtifact],
    run_id: &str,
    run_name: &str,
    output_dir: &Path,
    project_root: &Path,
    manifest_yaml_path: Option<&Path>,
    git_info_path: Option<&Path>,
    metadata_path: Option<&Path>,
) -> AppResult<PathBuf> {
    fs::create_dir_all(output_dir).map_err(|e| format!("create output dir failed: {e}"))?;
    let archive_path = output_dir.join(format!("{run_id}.tar.gz"));

    let output = File::create(&archive_path).map_err(|e| format!("create archive failed: {e}"))?;
    let encoder = GzEncoder::new(output, Compression::new(manifest.archive.compression_level));
    let mut tar = Builder::new(encoder);

    let mut used: HashSet<String> = HashSet::new();
    for artifact in artifacts.iter().filter(|x| x.status == ArtifactStatus::Included) {
        for (source, dest_name) in
            iter_artifact_members(artifact, run_name, project_root, &mut used)?
        {
            tar.append_path_with_name(&source, &dest_name)
                .map_err(|e| format!("add file to archive failed: {e}"))?;
        }
    }

    if manifest.archive.include_manifest {
        if let Some(manifest_path) = manifest_yaml_path {
            tar.append_path_with_name(manifest_path, format!("{run_name}/manifest.yaml"))
                .map_err(|e| format!("append manifest failed: {e}"))?;
        }
    }
    if manifest.archive.include_git_info {
        if let Some(git_path) = git_info_path {
            tar.append_path_with_name(git_path, format!("{run_name}/git_info.json"))
                .map_err(|e| format!("append git_info failed: {e}"))?;
        }
    }
    if manifest.archive.include_metadata {
        if let Some(meta_path) = metadata_path {
            tar.append_path_with_name(meta_path, format!("{run_name}/metadata.json"))
                .map_err(|e| format!("append metadata failed: {e}"))?;
        }
    }

    let encoder = tar
        .into_inner()
        .map_err(|e| format!("close tar builder failed: {e}"))?;
    encoder
        .finish()
        .map_err(|e| format!("close gzip encoder failed: {e}"))?;

    Ok(archive_path)
}

fn iter_artifact_members(
    artifact: &CollectedArtifact,
    run_name: &str,
    project_root: &Path,
    used: &mut HashSet<String>,
) -> AppResult<Vec<(PathBuf, String)>> {
    let mut out = Vec::new();

    match artifact.artifact_type.as_str() {
        "file" => {
            if let Some(first) = artifact.matched_paths.first() {
                let source = PathBuf::from(first);
                let base = source.file_name().and_then(|v| v.to_str()).unwrap_or("file");
                let final_name = unique_name(base, used);
                let dest = Path::new(run_name)
                    .join("artifacts")
                    .join(&artifact.name)
                    .join(final_name);
                out.push((source, dest.to_string_lossy().to_string()));
            }
        }
        "directory" => {
            if let Some(root_path) = artifact.matched_paths.first() {
                let source = PathBuf::from(root_path);
                for entry in WalkDir::new(&source).into_iter().filter_map(Result::ok) {
                    if entry.file_type().is_dir() {
                        continue;
                    }
                    let path = entry.path().to_path_buf();
                    if should_skip_file(project_root, &source, &path) {
                        continue;
                    }
                    let rel = path.strip_prefix(&source).unwrap_or(&path);
                    let dest = Path::new(run_name)
                        .join("artifacts")
                        .join(&artifact.name)
                        .join(rel);
                    out.push((path, dest.to_string_lossy().to_string()));
                }
            }
        }
        "glob" => {
            for path_text in &artifact.matched_paths {
                let p = PathBuf::from(path_text);
                if p.is_file() {
                    let base = p.file_name().and_then(|v| v.to_str()).unwrap_or("file");
                    let final_name = unique_name(base, used);
                    let dest = Path::new(run_name)
                        .join("artifacts")
                        .join(&artifact.name)
                        .join(final_name);
                    out.push((p, dest.to_string_lossy().to_string()));
                } else if p.is_dir() {
                    let group = unique_name(
                        p.file_name().and_then(|v| v.to_str()).unwrap_or("dir"),
                        used,
                    );
                    for entry in WalkDir::new(&p).into_iter().filter_map(Result::ok) {
                        if entry.file_type().is_dir() {
                            continue;
                        }
                        let path = entry.path().to_path_buf();
                        if should_skip_file(project_root, &p, &path) {
                            continue;
                        }
                        let rel = path.strip_prefix(&p).unwrap_or(&path);
                        let dest = Path::new(run_name)
                            .join("artifacts")
                            .join(&artifact.name)
                            .join(&group)
                            .join(rel);
                        out.push((path, dest.to_string_lossy().to_string()));
                    }
                }
            }
        }
        _ => {}
    }

    Ok(out)
}

fn should_skip_file(project_root: &Path, root: &Path, path: &Path) -> bool {
    let rel_root = match root.strip_prefix(project_root) {
        Ok(r) => r,
        Err(_) => return false,
    };
    let rel = match path.strip_prefix(project_root) {
        Ok(r) => r,
        Err(_) => return false,
    };
    if rel_root
        .components()
        .map(|c| c.as_os_str().to_string_lossy())
        .any(|part| DEFAULT_EXCLUDED_DIRS.iter().any(|d| part == *d))
    {
        return true;
    }
    for part in rel.components() {
        let part = part.as_os_str().to_string_lossy();
        if DEFAULT_EXCLUDED_DIRS.iter().any(|d| part == *d) {
            return true;
        }
    }
    false
}

fn unique_name(base: &str, used: &mut HashSet<String>) -> String {
    if !used.contains(base) {
        used.insert(base.to_string());
        return base.to_string();
    }

    let stem = Path::new(base)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or(base);
    let suffix = Path::new(base).extension().and_then(|s| s.to_str()).unwrap_or("");

    let mut idx = 1u32;
    loop {
        let candidate = if suffix.is_empty() {
            format!("{stem}_{idx}")
        } else {
            format!("{stem}_{idx}.{suffix}")
        };
        if !used.contains(&candidate) {
            used.insert(candidate.clone());
            return candidate;
        }
        idx += 1;
    }
}

pub fn write_manifest_snapshot(content: &str, path: &Path) -> AppResult<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("create manifest parent failed: {e}"))?;
    }
    fs::write(path, content).map_err(|e| format!("write manifest snapshot failed: {e}"))
}
